src/archive.rs

```diff
@@ -0,0 +1,154 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Tools for creating and inserting into tarballs.
+
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use camino::Utf8Path;
+use flate2::write::GzEncoder;
+use std::convert::TryInto;
+use std::fs::{File, OpenOptions};
+use tar::Builder;
+
+#[async_trait]
+pub trait AsyncAppendFile {
+    async fn append_file_async<P>(&mut self, path: P, file: &mut File) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send;
+
+    async fn append_path_with_name_async<P, N>(&mut self, path: P, name: N) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send,
+        N: AsRef<Utf8Path> + Send;
+
+    async fn append_dir_all_async<P, Q>(&mut self, path: P, src_path: Q) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send,
+        Q: AsRef<Utf8Path> + Send;
+}
+
+#[async_trait]
+impl<W: Encoder> AsyncAppendFile for Builder<W> {
+    async fn append_file_async<P>(&mut self, path: P, file: &mut File) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send,
+    {
+        tokio::task::block_in_place(move || self.append_file(path.as_ref(), file))
```

Comment:
I haven't used `block_in_place` much, I think I have an idea of why you're using it over `spawn_blocking`: so that it doesn't have to be `'static` is that correct?

Reply:
Yeah, that's one of the reasons -- the other is the access to `self`, which isn't `'static`. It would probably need to be clone-able to work with `spawn_blocking` here.